sklearn/metrics/pairwise.py

```diff
@@ -585,6 +597,273 @@ def cosine_distances(X, Y=None):
+    X_cat = X[:, categorical_features]
+
+    # Numerical columns
+    X_num = X[:, np.logical_not(categorical_features)]
+
+    Y_cat = None
+    Y_num = None
+    if not y_none:
+        Y_cat = Y[:, categorical_features]
+        Y_num = Y[:, np.logical_not(categorical_features)]
+    else:
+        Y_cat = X_cat
+        Y_num = X_num
+
+    # Calculates the min and max values to normalize the input values
+    # as proposed by the Gower's paper
+    _, num_cols = X_num.shape
+    max_of_numeric = np.zeros(num_cols)
+    abs_min_of_numeric = np.zeros(num_cols)
+
+    for col in range(num_cols):
+        col_array = X_num[:, col].astype(np.float32)
+        max = np.nanmax(col_array)
+        min = np.nanmin(col_array)
+
+        if np.isnan(max):
+            max = 0.0
+        if np.isnan(min):
+            min = 0.0
+
+        if not y_none:
+            col_array = Y_num[:, col].astype(np.float32)
+            min = np.nanmin([min, np.nanmin(col_array)])
+            max = np.nanmax([max, np.nanmax(col_array)])
+
+        if min < 0:
+            abs_min_of_numeric[col] = np.abs(min)
+            max = np.abs(min) + max
+            min = np.abs(min) + min
+        else:
+            abs_min_of_numeric[col] = 0
+            max = max - min
+            min = 0
+
+        max_of_numeric[col] = max
+
+    # This is to normalize the numeric values between 0 and 1.
+    X_num = np.divide(abs_min_of_numeric + X_num,
+                      max_of_numeric, out=np.zeros_like(X_num),
+                      where=max_of_numeric != 0)
+
+    if not y_none:
+        Y_num = np.divide(abs_min_of_numeric + Y_num,
+                          max_of_numeric, out=np.zeros_like(Y_num),
+                          where=max_of_numeric != 0)
+    else:
+        Y_num = X_num
+
+    y_n_rows, _ = Y.shape
+
+    D = np.zeros((n_rows, y_n_rows), dtype=np.float32)
+
+    Y_cat = None
+    Y_num = None
+
+    if not y_none:
+        Y_cat = Y[:, categorical_features]
+        Y_num = Y[:, np.logical_not(categorical_features)]
+        # This is to normalize the numeric values between 0 and 1.
+        Y_num = np.divide(Y_num, max_of_numeric, out=np.zeros_like(Y_num),
+                          where=max_of_numeric != 0)
+    else:
+        Y_cat = X_cat
+        Y_num = X_num
+
+    # This is to ignore FutureWarnings, as numpy will change
+    # how object arrays are compared in _gower_distance_row,
+    # as there is no other proposal to execute this efficiently
+    # without to avoid this warning.
+    warnings.filterwarnings("ignore",
+                            "numpy equal will not check object identity" +
+                            " in the future")
+    for i in range(n_rows):
+        j_start = i
+
+        # for non square results
+        if n_rows != y_n_rows or not y_none:
+            j_start = 0
+
+        Y_cat[j_start:n_rows, :]
+        Y_num[j_start:n_rows, :]
+        result = _gower_distance_row(X_cat[i, :], X_num[i, :],
```

Comment:
Let's inline this. I don't think it's helpful to put this in a different function